Add builder image selector in card layout and import flow refactor

rohitkrai03 opened the pull request:
Related JIRA story - https://jira.coreos.com/browse/ODC-690

This PR
- Adds a new Builder Image Selector which is based on Card layout.
- Also adds a builder image version selector which show the tag selected for builder image.
- Based on the builder image and tag selected, builder image information will be showed.
- Now auto detected image will be shown as recommended image with a green check icon.
- User will be able to select a different builder image than recommended image.

![ezgif com-video-to-gif (1)](https://user-images.githubusercontent.com/6041994/57667509-60b53f00-7621-11e9-8877-b9775589d4b0.gif)

rohitkrai03 committed:
Add builder image selector in card layout and import flow refactor

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageSelector.scss:
```diff
@@ -0,0 +1,9 @@
+  flex-flow: wrap;
+  background: var(--pf-global--Color--light-200);
+  margin-top: 10px;
+  margin-bottom: 10px;
```
Remove the margins. The spacing is applied through form groups.

Add extra padding to the box to give even spacing around the outside to equal the spacing between items:
![image](https://user-images.githubusercontent.com/14068621/57636759-ccdb6880-7577-11e9-836a-7b6218db3295.png)

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageCard.scss:
```diff
@@ -0,0 +1,39 @@
+.odc-builder-image-selector__card {
```
```suggestion
.odc-builder-image-card {
```

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageCard.scss:
```diff
@@ -0,0 +1,39 @@
+  margin: 8px;
+  text-align: center;
+
+  &-icon {
```
```suggestion
  &__icon {
```
Same pattern for all the other element selectors.

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageCard.scss:
```diff
@@ -0,0 +1,39 @@
+  }
+
+  &-selected {
+    border: var(--pf-global--active-color--100) 2px solid;
```
Use `outline` instead of border to prevent the card from bouncing in size.
Modifiers should use the format `is-<modifier>`.

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageCard.scss:
```diff
@@ -0,0 +1,39 @@
+
+  &-recommended {
+    position: absolute;
+    color: #3f9c35;
```
Use a PF color variable.

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageCard.scss:
```diff
@@ -0,0 +1,39 @@
+  }
+
+  &:hover {
+    border: var(--pf-global--active-color--100) 2px solid;
```
Use `outline` instead of `border` and join this selector with the selected one above.

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageCard.scss:
```diff
@@ -0,0 +1,39 @@
+  &-title {
+    font-size: 12px;
+    font-weight: 500;
+    line-height: 18px;
```
Use patternfly variables.
I think you need `font-weight: 700` to make it bold like the design mock up.

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageCard.scss:
```diff
@@ -0,0 +1,39 @@
+  position: relative;
+  height: 110px;
+  width: 132px;
+  margin: 8px;
```
PF uses size increments of 8px. Could you update your sizes accordingly.

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageCard.tsx:
```diff
@@ -0,0 +1,34 @@
+    <Card
+      className={`odc-builder-image-selector__card ${
+        selected ? 'odc-builder-image-selector__card-selected' : ''
+      }`}
```
Use the library `classnames` to create class strings.

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageCard.tsx:
```diff
@@ -0,0 +1,34 @@
+import { OkIcon } from '@patternfly/react-icons';
+import './BuilderImageCard.scss';
+
+const BuilderImageCard = ({ image, selected, recommended, onChange }) => {
```
Need prop types definition.

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageCard.tsx:
```diff
@@ -0,0 +1,34 @@
+
+const BuilderImageCard = ({ image, selected, recommended, onChange }) => {
+  return (
+    <Card
```
```suggestion
    <Card component="button" aria-label={image.title}
```
You'll have to adjust the styles a bit and fix the removal of the focus outline that console removes in `_base.scss` for whatever reason. But if you do this then you can tab through the buttons instead of just using the mouse.

rohitkrai03 replied:
Done.

christianvogt replied:
To fix the form submit, you also need to set `type="button"`

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageSelector.tsx:
```diff
@@ -0,0 +1,65 @@
+  selectedImage,
+  recommendedImage,
+  isBuilderImageDetected,
+  gitUrlValidationStatus,
```
Do not bring the definition of `gitUrlValidationStatus` into this component. If it is just being used to show a loading image, make a specific prop for that purpose.

christianvogt commented on frontend/public/extend/devconsole/utils/imagestream-utils.ts:
```diff
@@ -44,3 +67,41 @@ export const getPorts = (imageStreamImage: object): ContainerPort[] => {
+      'annotations',
+      'openshift.io/display-name',
+    ]);
+    const title = displayName && displayName.length < 14 ? displayName : prettifyName(name);
```
Why use `displayName` only if the length is < 14?

rohitkrai03 replied:
Because of the card size we cannot show display name with more than 14 characters. So instead of truncating it with `...` i am using spec name after prettifying it which is generally shorter.

christianvogt replied:
@rohitkrai03 is this common to have longer names? Are they single words or multiple words that can wrap?
Did you bring this up with UX?

rohitkrai03 replied:
@christianvogt  Yes, i did bring this up with UX - https://coreos.slack.com/archives/CHC2R6AGG/p1557398418064500. We have names like `Nginx HTTP server and a reverse proxy (nginx)`. The response i got from UX was to truncate the name and add `...` but it doesn't seem right to me. That is why i chose to display spec name which is shorter.

christianvogt replied:
Thanks for the clarification. We go with this unless told otherwise :)

christianvogt commented on frontend/public/extend/devconsole/utils/imagestream-utils.ts:
```diff
@@ -44,3 +67,41 @@ export const getPorts = (imageStreamImage: object): ContainerPort[] => {
+): NormalizedBuilderImages => {
+  const builderImageStreams = imageStreams.filter((imageStream) => isBuilder(imageStream));
+  const builderImages = {};
+  builderImageStreams.map((imageStream) => {
```
`map` => `each`
or use `reduce`

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageTagSelector.tsx:
```diff
@@ -0,0 +1,55 @@
+import ImageStreamInfo from '../source-to-image/ImageStreamInfo';
+
+export interface BuilderImageTagSelectorProps {
+  imageTags: any[];
```
Can you use a type that's not `any`?

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageSelector.tsx:
```diff
@@ -0,0 +1,65 @@
+}) => {
+  if (!builderImages) {
+    return <LoadingInline />;
+  }
```
Use a prop `loading` to indicate whether the loading component should be shown. Let the parent set `loading={!builderImages}`

rohitkrai03 replied:
Done

christianvogt replied:
Looks odd to not show the label while the data is loading. The form group should still be present.

christianvogt commented on frontend/public/extend/devconsole/components/ImportFlowForm/ImportFlowForm.tsx:
```diff
@@ -487,100 +499,81 @@ export class ImportFlowForm extends React.Component<Props, State> {
+              data-test-id="import-name"
+            />
+            <HelpBlock>
+              {nameError ? nameError : 'Identifies the resources created for this application'}
```
```suggestion
              {nameError || 'Identifies the resources created for this application'}
```

christianvogt commented:
Bug: When the user has already chosen a builder image, the version and description is shown. If the user then goes and changes the git URL, detection is re-run and then the builder image version description disappears.

christianvogt commented:
@rohitkrai03 as seen in your screenshot, the builder image selector expands beyond the bounds of the form. It should be the same width as the rest of the form.

rohitkrai03 committed:
Fix CSS class names, some styles fixes, added type for tag and other fixes.

rohitkrai03 commented:
@christianvogt I've updated the PR with fixes according to review comments. Please review again. Also fixed the bug you mentioned.
About the size of builder image selector, i've done this based on the designs in https://redhat.invisionapp.com/share/G2RH3HM4HMT#/screens/358681874. It seemed odd to me as well that only the card layout is supposed to be full width and i had to actually add extra styling to make this happen. I am happy to make it same as other form elements if you want.

invincibleJai commented:
@rohitkrai03 shall we make use of https://github.com/typicode/react-fake-props , this will eliminate having props in code

christianvogt commented:
@invincibleJai with typescript the prop types are only a compile time artifact. For plain JS, the prop types are typically removed via a babel plugin. Not sure if console uses it but regardless we use typescript.

invincibleJai commented:
thanks, @christianvogt I was thinking `react-fake-props` can eliminate having props from test as in  https://github.com/talamer/console/blob/master-next/frontend/public/extend/devconsole/components/ImportFlowForm/__tests__/ImportFlowForm.spec.tsx#L12-L79  and maybe look cleaner.

christianvogt commented:
@invincibleJai ah sorry. I didn't pick up exactly on what it's intent was.

christianvogt commented:
While the builder image is loading we see this; missing the label:
![image](https://user-images.githubusercontent.com/14068621/57723904-adb40800-7657-11e9-8daa-1a1322137548.png)

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageCard.scss:
```diff
@@ -0,0 +1,37 @@
+.odc-builder-image-card {
```
add `cursor: pointer` if the supplied `component` to the `Card` is not a semantic HTML element that provides the `pointer` cursor by default.

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageCard.scss:
```diff
@@ -0,0 +1,37 @@
+
+  &__recommended {
+    position: absolute;
+    color: var(--pf-global--success-color--200);
```
![image](https://user-images.githubusercontent.com/14068621/57724486-e1dbf880-7658-11e9-8c26-c28653b36082.png)

This looks dark. Maybe use the `100` variant? Ask UX.
Also the recommended text above doesn't have the same color for the star.
![image](https://user-images.githubusercontent.com/14068621/57724450-d1c41900-7658-11e9-8f06-b9a3fbecb4b3.png)

christianvogt replied:
Not going to hold up the PR on a color, but both stars should at least be the same color.

rohitkrai03 replied:
Well, i used lighter colour at first but that doesn't catch user's attention easily. So switched to darker ones.
<img width="347" alt="Screenshot 2019-05-15 at 1 15 16 AM" src="https://user-images.githubusercontent.com/6041994/57727416-17abcc00-76af-11e9-8fe1-498570e606b8.png">
<img width="144" alt="Screenshot 2019-05-15 at 1 01 12 AM" src="https://user-images.githubusercontent.com/6041994/57727430-1e3a4380-76af-11e9-8504-0e05c1145867.png">

If it looks ok to you then i am fine.

christianvogt replied:
I wasn't sure if we wanted to replace the validation icons as well. I don't really don't care at this point :P

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageCard.tsx:
```diff
@@ -0,0 +1,39 @@
+}) => {
+  const classes = classNames('odc-builder-image-card', { 'is-selected': selected });
+  return (
+    <Card className={classes} onClick={() => onChange(image.name)}>
```
By default a `Card` is an `article` which doesn't fit semantically for this usage.

rohitkrai03 replied:
Fixed it. Now the buttons are also accessible through tabs.

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageSelector.tsx:
```diff
@@ -0,0 +1,72 @@
+  }
+
+  let showDetectBuildToolStatus;
+  if (isGitUrlValidated && !recommendedImage && !builderImageError) {
```
Replace with a prop `loadingRecommendedImage` and move the logic for setting the value `true|false` to the parent.

rohitkrai03 replied:
Done

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageTagSelector.tsx:
```diff
@@ -0,0 +1,56 @@
+          dropDownClassName="dropdown--full-width"
+        />
+      </FormGroup>
+      {selectedImageTag && (
```
Use `imageTag` since it's the data that is passed to the child component.
```suggestion
      {imageTag && (
```

christianvogt commented on frontend/public/extend/devconsole/components/source-to-image/SourceToImage.tsx:
```diff
@@ -212,6 +220,12 @@ class BuildSource extends React.Component<
+      tag,
+      ['annotations', 'openshift.io/display-name'],
+      imageStream.metadata.name,
+    );
```
Same thing done twice now. Looks like a good candidate for a utility.

rohitkrai03 replied:
Done. Added a utility function that finds the tag and display name and returns both.

rohitkrai03 committed:
Show loading animations along with label for builder image

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageCard.tsx:
```diff
@@ -0,0 +1,39 @@
+      </CardBody>
+      {recommended && (
+        <span className="odc-builder-image-card__recommended">
+          <i className="fa fa-star" aria-hidden="true" />
```
use patternfly react-icons

rohitkrai03 replied:
The new Patternfly 4 docs didn't have star icon listed - https://www.patternfly.org/v4/design-guidelines/styles/icons that's why i used font awesome. Just now looked at http://patternfly-react.surge.sh/patternfly-4/icons/Icons/ and found star. Will use it.

rohitkrai03 commented:
@christianvogt Fixed. Now the loading animation is shown along with label.

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageSelector.tsx:
```diff
@@ -0,0 +1,73 @@
+      {isBuilderImageDetected && (
+        <HelpBlock>
+          (Recommended builder images are represented by{' '}
+          <i className="fa fa-star" aria-hidden="true" /> icon)
```
use patternfly react-icons

rohitkrai03 replied:
Done

rohitkrai03 replied:
done

rohitkrai03 committed:
Made cards semantically correct, fixed icon color and other css issues, created utility for getting tag data.

rohitkrai03 commented:
@christianvogt Updated PR with fixes according to review comments.

vikram-raj reviewed (changes requested):
When I change the autodetected builder image by selecting the other card then it submits the form which should not happen.
![Peek 2019-05-15 08-23](https://user-images.githubusercontent.com/2561818/57745563-4ba4e280-76eb-11e9-8709-47b4e7626e6c.gif)

vikram-raj commented:
Card shadows look darker than the actual design. 

Current implementation
![Screenshot from 2019-05-15 08-29-00](https://user-images.githubusercontent.com/2561818/57745722-dd145480-76eb-11e9-8eca-a22e3b6be1c7.png)

Actual design
![Screenshot from 2019-05-15 08-32-24](https://user-images.githubusercontent.com/2561818/57745759-ff0dd700-76eb-11e9-9213-11bc4c805e29.png)

christianvogt commented:
The buttons look like this for me now:
![image](https://user-images.githubusercontent.com/14068621/57788176-a26ae600-7704-11e9-84ee-ca83110c6fac.png)

You need to remove the border that's applied to buttons.

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageCard.scss:
```diff
@@ -0,0 +1,36 @@
+
+  &.is-selected,
+  &:hover {
+    outline: var(--pf-global--active-color--100) 3px solid;
```
If continuing with using outline:
```
  &.is-selected,
  &:hover,
  &:focus,
  &:active {
    outline: var(--pf-global--active-color--100) 3px solid;
    outline-offset: -3px;
  }
```

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageCard.scss:
```diff
@@ -0,0 +1,36 @@
+.odc-builder-image-card {
```
Add `outline: 0;` if using outline for your border. Otherwise set a transparent border equal to the size of the hover / active / focus border.

Also need border of 0 if continuing to use outline in order to negate the default look of buttons in browsers.

christianvogt commented on frontend/public/extend/devconsole/components/import/BuilderImageSelector.scss:
```diff
@@ -0,0 +1,12 @@
+  padding: 4px;
+  cursor: pointer;
+
+  button:focus {
```
Remove this. If needed it should be part of the card itself and not here.

rohitkrai03 committed:
Fixed issue with auto submit on card selection and css issues.

rohitkrai03 commented:
@christianvogt @vikram-raj Updated PR with fixes to above issues.
